ipn/ipnlocal/local.go

```diff
@@ -6203,3 +6205,164 @@ func mayDeref[T any](p *T) (v T) {
+
+// SuggestExitNode returns a tailcfg.StableNodeID, name, and tailcfg.Location of a suggested exit node given the local backend's netmap and last report.
+// Non-mullvad nodes are prioritized before suggesting Mullvad nodes.
+// We look at non-mullvad nodes region latency values, and if there are none then we choose the geographic closest Mullvad node to the self node's preferred DERP region.
+func (b *LocalBackend) SuggestExitNode() (response apitype.ExitNodeSuggestionResponse, err error) {
+	b.mu.Lock()
+	lastReport := b.MagicConn().GetLastNetcheckReport(b.ctx)
+	netMap := b.netMap
+	b.mu.Unlock()
+	seed := time.Now().UnixNano()
+	return suggestExitNode(lastReport, netMap, seed)
+}
+
+func suggestExitNode(lastReport *netcheck.Report, netMap *netmap.NetworkMap, seed int64) (response apitype.ExitNodeSuggestionResponse, err error) {
+	var preferredExitNodeID tailcfg.StableNodeID
+	var preferredExitNodeName string
+	var preferredExitNodeLocation tailcfg.Location
+	if lastReport.PreferredDERP == 0 {
+		return apitype.ExitNodeSuggestionResponse{ID: preferredExitNodeID, Name: "", Location: tailcfg.Location{}}, ErrNoPreferredDerp
+	}
+	var candidates []tailcfg.NodeView
+	peers := netMap.Peers
+	for _, peer := range peers {
+		if peer.CapMap().Has(tailcfg.NodeAttrSuggestExitNode) {
+			if online := peer.Online(); online != nil && !*online {
+				continue
+			}
+			if tsaddr.ContainsExitRoutes(peer.AllowedIPs()) {
+				candidates = append(candidates, peer)
+			}
+		}
+	}
+	if len(candidates) == 0 {
+		return apitype.ExitNodeSuggestionResponse{ID: preferredExitNodeID, Name: "", Location: tailcfg.Location{}}, nil
+	}
+	if len(candidates) == 1 {
+		hi := candidates[0].Hostinfo()
+		if hi.Valid() {
+			if loc := hi.Location(); loc != nil {
+				preferredExitNodeLocation = *loc
+			}
+		}
+		return apitype.ExitNodeSuggestionResponse{ID: candidates[0].StableID(), Name: candidates[0].Name(), Location: preferredExitNodeLocation}, nil
+	}
+
+	candidatesByRegion := make(map[int][]tailcfg.NodeView, len(netMap.DERPMap.Regions))
+	candidateDistanceMap := make(map[float64][]tailcfg.NodeView, len(candidates))
+	preferredDerp := netMap.DERPMap.Regions[lastReport.PreferredDERP]
+	regions := make([]int, 0, len(netMap.DERPMap.Regions))
+	var minDistance float64 = math.MaxFloat64
+	if preferredDerp == nil {
+		return apitype.ExitNodeSuggestionResponse{ID: preferredExitNodeID, Name: "", Location: tailcfg.Location{}}, ErrNoPreferredDerp
+	}
+	for _, candidate := range candidates {
+		if candidate.Valid() {
+			if candidate.DERP() != "" {
+				ipp, err := netip.ParseAddrPort(candidate.DERP())
+				if err != nil {
+					continue
+				}
+				if ipp.Addr() == tailcfg.DerpMagicIPAddr {
+					regionID := int(ipp.Port())
+					if !slices.Contains(regions, regionID) {
+						regions = append(regions, regionID)
+					}
+					candidatesByRegion[regionID] = append(candidatesByRegion[regionID], candidate)
+				}
+			} else {
```

Comment:
```suggestion
				continue
			}
			// This candidate does not have a DERP home
			if len(candidatesByRegion) > 0 {
				// Since a candidate exists that does have a DERP home, skip this candidate. We never select
				// a candidate without a DERP home if there is a candidate available with a DERP home.
			}
			// Use geographic distance from our DERP home to estimate how good this candidate is.
```